File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    
    public int distance(int loc, int num){
        int dis = 0;
        if(loc%3==0){
            dis++;
            dis += Math.abs(num+1-loc)/3;
        }else if(loc%3==1){
            dis++;
            dis += Math.abs(num-1-loc)/3;
        }else{
            dis += Math.abs(num/3-loc/3);
        }
        return dis;
    }
    
    public String solution(int[] numbers, String hand) {
        StringBuffer answer = new StringBuffer("");
        int right=12, left=10;
        int rcnt=0, lcnt=0;
        String use = "";
        for(int num : numbers){
            if(num==0) num = 11;
            if(num%3==1){
                use = "L";
            }else if(num%3==0){
                use = "R";
            }else{
                rcnt = distance(right, num);
                lcnt = distance(left, num);
                
                if(rcnt>lcnt){
                    use = "L";
                }else if(rcnt<lcnt){
                    use = "R";
                }else{
                    use = hand.equals("right") ? "R" : "L" ;
                }
            }
            
            answer.append(use);
            if(use.equals("R")) 
                right = num;
            else 
                left = num;
        }
        return answer.toString();
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    var L = [3, 0]
    var R = [3, 2]
    
    return numbers.map(n => {
        if(n%3 === 1) {
            L = [Math.floor(n/3), 0];
            return 'L';
        }
        if(n%3 === 0 && n !== 0) {
            R = [n/3 - 1, 2]
            return 'R';
        }
        
        var row = Math.floor(n === 0 ? 3 : n/3);
        var l_diff = Math.abs(row - L[0]) + Math.abs(L[1] - 1);
        var r_diff = Math.abs(row - R[0]) + Math.abs(R[1] - 1);
        
        if(l_diff === r_diff) {
            if(hand === 'left'){
                L = [row, 1];
                return 'L'
            }
            else {
                R = [row, 1];
                return 'R'
            }
        }
        
        if(l_diff < r_diff){
            L = [row, 1];
            return 'L'
        }
        else {
            R = [row, 1];
            return 'R'
        }
    }).join('')
}